sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnosticsContext.java

```diff
@@ -556,10 +581,233 @@ public String toJson() {
+                        events.add(new CosmosDiagnosticsRequestEvent(e.getStartTime(), e.getDuration(), e.getName()));
+                    }
+                });
+            }
+
+            Duration backendLatency = null;
+            if (resultDiagnostics.getBackendLatencyInMs() != null) {
+                backendLatency = Duration.ofNanos((long)(resultDiagnostics.getBackendLatencyInMs() * 1000000d));
+            }
+
+            CosmosDiagnosticsRequestInfo info = new CosmosDiagnosticsRequestInfo(
+                activityId,
+                partitionId,
+                pkRangeId,
+                responseStats.getRequestResourceType() + ":" + responseStats.getRequestOperationType(),
+                requestStats.getRequestStartTimeUTC(),
+                responseStats.getDuration(),
+                backendLatency,
+                requestCharge,
+                responsePayloadLength,
+                statusCode,
+                subStatusCode,
+                events
+            );
+
+            requestInfo.add(info);
+        }
+    }
+
+    private void addRequestInfoForAddressResolution(
+        ClientSideRequestStatistics requestStats,
+        ArrayList<CosmosDiagnosticsRequestInfo> requestInfo,
+        Map<String, ClientSideRequestStatistics.AddressResolutionStatistics> addressResolutionStatisticsMap
+    ) {
+        if (addressResolutionStatisticsMap == null || addressResolutionStatisticsMap.size() == 0) {
+            return;
+        }
+
+        for (ClientSideRequestStatistics.AddressResolutionStatistics addressResolutionStatistics
+            : addressResolutionStatisticsMap.values()) {
+
+            if (addressResolutionStatistics.isInflightRequest() ||
+                addressResolutionStatistics.getEndTimeUTC() == null) {
+
+                // skipping inflight or failed address resolution statistics
+                // capturing error count etc. won't make sense here - request diagnostic
+                // logs are the right way to debug those - not metrics
+                continue;
+            }
+
+            Duration latency = Duration.between(
+                addressResolutionStatistics.getStartTimeUTC(),
+                addressResolutionStatistics.getEndTimeUTC());
+
+            StringBuilder sb = new StringBuilder();
+            sb.append("AddressResolution|");
+            sb.append(addressResolutionStatistics.getTargetEndpoint());
+            sb.append("|");
+            if (addressResolutionStatistics.isForceRefresh()) {
+                sb.append("1|");
+            } else {
+                sb.append("0|");
+            }
+
+            if (addressResolutionStatistics.isForceCollectionRoutingMapRefresh()) {
+                sb.append("1");
+            } else {
+                sb.append("0");
+            }
+
+            CosmosDiagnosticsRequestInfo info = new CosmosDiagnosticsRequestInfo(
+                requestStats.getActivityId(),
```

Comment:
addressResolution use a different activityId vs request.ActivityId, should we use the addressResolutionRequest's activityId? (the key from addressResolutionStatisticsMap). or we are purposely using the same activityId for grouping purpose?

Reply:
Fixed in next iteration